csharp/ql/lib/semmle/code/csharp/security/dataflow/ZipSlipQuery.qll

```diff
@@ -3,147 +3,360 @@
+          or
+          // Or if the function returns the resultant of the guard call
+          DataFlow::localFlow(DataFlow::exprNode(g), DataFlow::exprNode(ret.getExpr()))
+        )
+      )
+    )
   }
 }
 
-/** An access to the `FullName` property of a `ZipArchiveEntry`. */
-class ArchiveFullNameSource extends Source {
-  ArchiveFullNameSource() {
-    exists(PropertyAccess pa | this.asExpr() = pa |
+class IndirectOverloadedWrapperSantizierMethod extends AbstractWrapperSanitizerMethod {
+  /**
+   * An IndirectOverloadedWrapperSanitizerMethod is a Method in which simply wraps /another/ wrapper.class
+   *
+   * Usually this will look like the following stanza:
+   * boolean someWrapper(string s){
+   *  return someWrapper(s, true);
+   * }
+   */
+  IndirectOverloadedWrapperSantizierMethod() {
+    forex(ReturnStmt ret | ret.getEnclosingCallable() = this |
+      ret.getExpr().(BoolLiteral).getBoolValue() = false
+      or
+      exists(ZipSlipGuard g |
+        // If the parameter flows directly to SanitizerMethodCall, and the resultant is returned
+        DataFlow::localFlow(DataFlow::parameterNode(paramFilename),
+          DataFlow::exprNode(g.getFilePathArgument())) and
+        DataFlow::localFlow(DataFlow::exprNode(g), DataFlow::exprNode(ret.getExpr()))
+      )
+    )
+  }
+}
+
+class WrapperSanitizerMethodCall extends SanitizerMethodCall {
+  // A Wrapped Sanitizer Method call (some function that is equally or more restrictive than our root sanitizer)
+  AbstractWrapperSanitizerMethod wrapperMethod;
+
+  WrapperSanitizerMethodCall() {
+    exists(AbstractWrapperSanitizerMethod sm |
+      this.getTarget() = sm and
+      wrapperMethod = sm
+    )
+  }
+
+  override Expr getFilePathArgument() {
+    result = this.getArgument(wrapperMethod.paramFilePath().getIndex())
+  }
+}
+
+private predicate wrapperCheckGuard(Guard g, Expr e, AbstractValue v) {
+  // A given wrapper method call, with the filePathArgument as a sink, that returns 'true'
+  g instanceof WrapperSanitizerMethodCall and
+  g.(WrapperSanitizerMethodCall).getFilePathArgument() = e and
+  v.(AbstractValues::BooleanValue).getValue() = true
+}
+
+/**
+ * A sanitizer for unsafe zip extraction.
+ */
+abstract class Sanitizer extends DataFlow::ExprNode { }
+
+class WrapperCheckSanitizer extends Sanitizer {
+  // A Wrapped RootSanitizer that is an explicit subset of RootSanitizer
+  WrapperCheckSanitizer() { this = DataFlow::BarrierGuard<wrapperCheckGuard/3>::getABarrierNode() }
+}
+
+/**
+ * A data flow source for unsafe zip extraction.
+ */
+abstract class Source extends DataFlow::Node { }
+
+class ArchiveEntryFullName extends Source {
+  // Access to full name of the archive item
+  ArchiveEntryFullName() {
+    exists(PropertyAccess pa |
       pa.getTarget().getDeclaringType().hasQualifiedName("System.IO.Compression", "ZipArchiveEntry") and
-      pa.getTarget().getName() = "FullName"
+      pa.getTarget().getName() = "FullName" and
+      this = DataFlow::exprNode(pa)
```

Comment:
same as this.getExpr() = pa?

Reply:
Yes however I like the explicit DataFlow::exprNode() form.

Reply:
You've sold me actually, the more I stare at it the more DataFlow::exprNode() feels cleaner to me too.